src/app/clusters/icd-management-server/icd-management-server.cpp

```diff
@@ -69,61 +71,198 @@ CHIP_ERROR IcdManagementAttributeAccess::Read(const ConcreteReadAttributePath &
 
 CHIP_ERROR IcdManagementAttributeAccess::ReadRegisteredClients(EndpointId endpoint, AttributeValueEncoder & encoder)
 {
-    return encoder.EncodeEmptyList();
+    uint16_t supported_clients = 0;
+    VerifyOrReturnError(EMBER_ZCL_STATUS_SUCCESS == Attributes::ClientsSupportedPerFabric::Get(endpoint, &supported_clients),
+                        CHIP_ERROR_NOT_FOUND);
+
+    return encoder.EncodeList([supported_clients](const auto & subEncoder) -> CHIP_ERROR {
+        IcdMonitoringEntry e;
+        uint16_t i     = 0;
+        CHIP_ERROR err = CHIP_NO_ERROR;
+
+        const auto & fabricTable = Server::GetInstance().GetFabricTable();
+        for (const auto & fabricInfo : fabricTable)
+        {
+            PersistentStorageDelegate & storage = chip::Server::GetInstance().GetPersistentStorage();
+            IcdMonitoringTable table(&storage, fabricInfo.GetFabricIndex(), supported_clients);
+            while ((CHIP_NO_ERROR == err) && (i < table.Limit()))
+            {
+                err = table.Get(i++, e);
+                if (CHIP_NO_ERROR == err)
+                {
+                    ReturnErrorOnFailure(subEncoder.Encode(e));
+                }
+            }
+        }
+        return (CHIP_ERROR_NOT_FOUND == err) ? CHIP_NO_ERROR : err;
+    });
+}
+
+CHIP_ERROR CheckAdmin(chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath, bool & is_admin)
+{
+    RequestPath requestPath{ .cluster = commandPath.mClusterId, .endpoint = commandPath.mEndpointId };
+    CHIP_ERROR err = GetAccessControl().Check(commandObj->GetSubjectDescriptor(), requestPath, Privilege::kAdminister);
+    if (CHIP_NO_ERROR == err)
+    {
+        is_admin = true;
+    }
+    else if (CHIP_ERROR_ACCESS_DENIED == err)
+    {
+        is_admin = false;
+        err      = CHIP_NO_ERROR;
+    }
+    return err;
 }
 
+class IcdManagementFabricDelegate : public chip::FabricTable::Delegate
+{
+    void OnFabricRemoved(const FabricTable & fabricTable, FabricIndex fabricIndex) override
+    {
+        uint16_t supported_clients = 0;
+        if (EMBER_ZCL_STATUS_SUCCESS == Attributes::ClientsSupportedPerFabric::Get(kRootEndpointId, &supported_clients))
```

Comment:
It really bothers me that we leak the storage if this get fails.  Can we somehow clean it up without needing this supported_clients value?

Reply:
This function now falls back to UINT16_MAX, so the remove function will loop until no more entries are found.